### PR TITLE
Depth estimation dataset support + Midas redweb dataset

### PR DESCRIPTION
**Pull request by philmarchenko**

### CODE REVIEW

File: src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    ):
+        """ """
+        super().__init__(transforms=transforms)
+        self.data_dir = data_dir
+        self.images_dir = images_dir
+        self.targets_dir = targets_dir
+        self.image_extension = image_extension
+        self.target_extension = target_extension
+
+        self.pair_names = [name.split(".")[0] for name in os.listdir(os.path.join(self.data_dir, self.targets_dir))]
```
Comment (BloodAxe) [inline src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py]: A few issues I see here:
1) `sorted(os.listdir(...))` Just to stay safe. listdir is not guaranteed to return file names in same order.
2) We are not validating that os.listdir returns all images of target_extension.
3) We are not validating that present images has the same filenames (set are equal) as targets.

So a better approach would be to:
- get all files in images dir 
- all files in targets dir
- filter both of them according to `image_extension` and `target_extension`
- check that two sets are equal

File: src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    def __len__(self):
+        return len(self.pair_names)
+
+    def load_sample(self, index: int) -> DepthEstimationSample:
+        pair_name = self.pair_names[index]
+
+        image = Image.open(os.path.join(self.data_dir, self.images_dir, f"{pair_name}.{self.image_extension}")).convert("RGB")
```
Comment (BloodAxe) [inline src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py]: See https://github.com/hakuryuu96/super-gradients/pull/1/files#r1421770713
Once there are `self.image_filename_list` and `self.targets_filename_list` you don't need to do os.path.join at all and can do `Image.open(self.image_filename_list[i])`

File: src/super_gradients/training/samples/depth_estimation_sample.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    :param image:              Image of [H, W, (C if colorful)] shape.
+    :param depth_map:          Depth map of [H, W] shape.
+    """
+
+    __slots__ = ["image", "depth_map"]
+
+    image: Union[np.ndarray, torch.Tensor]
+    depth_map: Union[np.ndarray, torch.Tensor]
+
+    def __init__(self, image: Union[np.ndarray, torch.Tensor, Image], depth_map: Union[np.ndarray, torch.Tensor, Image]):
```
Comment (BloodAxe) [inline src/super_gradients/training/samples/depth_estimation_sample.py]: No need for torch.Tensor and especially for PIL.Image support here. In OD it was more for backward compatibility needs. Whenever possible let's stick to `np.ndarray` in the sample class.
The only place where we need to go to torch - is the collate function.

File: src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+        depth_map = Image.open(os.path.join(self.data_dir, self.targets_dir, f"{pair_name}.{self.target_extension}"))
```
Comment (BloodAxe) [inline src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py]: I think somewhere here we have to have the option to pass (or at least know) how the "invalid" areas are marked in the sample.
Invalid areas are the ones that are missing depth information. I imagine this is crucial information should not be lost as transformations should respect that and do not change those areas as well as metrics which should respect that.
So I have two ideas about this:
- Either introduce a convention that all non-finite values in mask are treated as ignored/invalid.
- Or introduce a field in DepthEstimationSample (`ignore_index` or something) that would have the value corresponding to ignored areas.

File: src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+        return DepthEstimationSample(image=image, depth_map=depth_map)
```
Comment (BloodAxe) [inline src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py:66]: Let's not use PIL containers at all and directly convert to numpy 🙏

File: src/super_gradients/training/samples/depth_estimation_sample.py (hunk 3 above)
Comment (BloodAxe) [inline src/super_gradients/training/samples/depth_estimation_sample.py]: I think we can show some love to the user and should check whether depth_map is indeed a [H,W] shape and there is no dummy channel dimension (And squeeze if it is there). This way we can ensure consistency between the documentation and the state of the sample.
Of course if the channel dimension > 1 that we should throw an error.

File: src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py (hunk 1 above)
Comment (philmarchenko) [reply src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py]: In the case of a ReDWeb dataset, I haven't seen any issues with file names, so an approach with storing only names doesn't require two additional lists with file paths and any kind of sorting. 

Do we foresee the case that users will somehow change the structure of the dataset offline?

File: src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py (hunk 4 above)
Comment (philmarchenko) [reply src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py]: Great idea! 

Although this dataset does not contain such information (and I don’t think this feature should be added here), it is worth taking into account in other datasets and perhaps even including it in the interface. I think that later when integrating datasets, this feature should be added.

File: src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py (hunk 5 above)
Comment (philmarchenko) [reply src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py:66]: I would happily not use PIL at all :)

Unfortunately, locally I faced a double behavior when tried to use np.arrays here. This appears when the list of transformations for a dataset is empty. If no transformations are applied, then np.array stays np.array (doesn't convert to PIL.Image) which is required for further steps. If someone has something like `x.size` in a code and expects to have an Image instance, they won't get `(d1, d2, ..., dn)` type of tuple.

Comment (philmarchenko) [reply src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py:66]: I would rather suggest removing support for PIL altogether, but not within the scope of this PR.

File: src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py (hunk 1 above)
Comment (BloodAxe) [reply src/super_gradients/training/datasets/depth_estimation_datasets/redweb_dataset.py]: > Do we foresee the case that users will somehow change the structure of the dataset offline?
Yes, we don't trust no one

- User can forget to put a correct dataset class (E.g instead say Hypersim put ReDWeb)
- User can have corrupted dataset or just copy some images by accident

The idea is to fail as early as possible.
If the dataset is not consistent and we are tryin to read non-existent image, with current design user will get an exception from the DataLoader at most which won't be telling much (Cannot read image bla-bla-bla.png)
When we check consistency in class c'tor - we can provide a more _meaningful_ message _earlier_.
